flow360/component/geometry.py

```diff
@@ -47,13 +54,132 @@ def _post_upload_heartbeat(info):
         time.sleep(HEARTBEAT_INTERVAL)
 
 
+class GeometryStatus(Enum):
+    """Status of geometry resource, the is_final method is overloaded"""
+
+    COMPLETED = "completed"
+    ERROR = "error"
+    DIVERGED = "diverged"
+    UPLOADED = "uploaded"
+    CASE_UPLOADED = "case_uploaded"
+    UPLOADING = "uploading"
+    RUNNING = "running"
+    PREPROCESSING = "preprocessing"
+    GENERATING = "generating"
+    PROCESSED = "processed"
+    STOPPED = "stopped"
+    DELETED = "deleted"
+    PENDING = "pending"
+    UNKNOWN = "unknown"
+
+    def is_final(self):
+        """
+        [Overloaded] Checks if status is final for geometry resource
+
+        Returns
+        -------
+        bool
+            True if status is final, False otherwise.
+        """
+        if self in [
+            GeometryStatus.COMPLETED,
+            GeometryStatus.DIVERGED,
+            GeometryStatus.ERROR,
+            GeometryStatus.PROCESSED,
+            GeometryStatus.DELETED,
+        ]:
+            return True
+        return False
+
+
 # pylint: disable=R0801
 class GeometryMeta(AssetMetaBaseModel):
     """
     GeometryMeta component
     """
 
     project_id: str = pd.Field(alias="projectId")
+    deleted: bool = pd.Field()
+    metadata: Optional[_GeometryMetadataModel] = pd.Field(None, const=False)
+    status: GeometryStatus = pd.Field()  # Overshadowing to ensure correct is_final() method
+
+    @classmethod
+    def _replace_metadata_in_given_geometry_meta(
+        cls, geometry_meta: GeometryMeta, new_metadata: _GeometryMetadataModel
+    ) -> GeometryMeta:
+        """
+        Replace the metadata inside the given geometry_meta with the
+        new_metadata and return the updated geometry_meta.
+        This circumvents the immutability of `AssetMetaBaseModel`.
+        """
+        geometry_meta_dict = geometry_meta.dict(by_alias=True)
+        geometry_meta_dict["metadata"] = new_metadata.dict()
+        return GeometryMeta.parse_obj(geometry_meta_dict)
```

Comment:
how about we just using `copy(update=)` ?
```python
gm = GeometryMeta(...)
gmm = _GeometryMetadataModel()
gm = gm.copy(update=dict(metadata=gmm))
```

Reply:
Right. I forgot this.

Reply:
Changed

Reply:
I think this class method can be now removed

Reply:
Removed